sdk/metric/aggregator/maxsumcount/msc.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package maxsumcount
+
+import (
+	"context"
+
+	"go.opentelemetry.io/api/core"
+	"go.opentelemetry.io/sdk/export"
+)
+
+type (
+	// Aggregator aggregates measure events, keeping only the max,
+	// sum, and count.
+	Aggregator struct {
+		live state
+		save state
+	}
+
+	state struct {
+		count core.Number
+		sum   core.Number
+		max   core.Number
+	}
+)
+
+var _ export.MetricAggregator = &Aggregator{}
+
+// New returns a new measure aggregator for computing max, sum, and count.
+func New() *Aggregator {
+	return &Aggregator{}
+}
+
+// SumAsInt64 returns the accumulated sum as an int64.
+func (c *Aggregator) SumAsInt64() int64 {
+	return c.save.sum.AsInt64()
+}
+
+// SumAsFloat64 returns the accumulated sum as an float64.
+func (c *Aggregator) SumAsFloat64() float64 {
+	return c.save.sum.AsFloat64()
+}
+
+// Count returns the accumulated count.
+func (c *Aggregator) Count() uint64 {
+	return c.save.count.AsUint64()
+}
+
+// MaxAsInt64 returns the accumulated max as an int64.
+func (c *Aggregator) MaxAsInt64() int64 {
+	return c.save.max.AsInt64()
+}
+
+// MaxAsFloat64 returns the accumulated max as an float64.
+func (c *Aggregator) MaxAsFloat64() float64 {
+	return c.save.max.AsFloat64()
+}
+
+// Collect saves the current value (atomically) and exports it.
+func (c *Aggregator) Collect(ctx context.Context, rec export.MetricRecord, exp export.MetricBatcher) {
+	desc := rec.Descriptor()
+	kind := desc.NumberKind()
+	zero := core.NewZeroNumber(kind)
+
+	// N.B. There is no atomic operation that can update all three
```

Comment:
We probably could solve it like with gauges:

```go
type Aggregator struct {
    state *state // or unsafe.Pointer
}

type state struct {
    live mscstate
    save mscstate
}

type mscstate struct {
    count core.Number
    sum core.Number
    max core.Number
}

// update
for {
    statePtr := (state*)(atomic.LoadPointer(c.state))
    copy := *statePtr
    copy.live.count += 1
    copy.live.sum += …
    copy.live.max = …
    if atomic.CompareAndSwapPointer(&c.state, statePtr, &copy) { break }
}

// collect
    statePtr := (state*)(atomic.LoadPointer(c.state))
    copy := *statePtr
    copy.save.count, copy.live.count = copy.live.count, 0
    copy.save.sum, copy.live.sum = copy.live.sum, 0
    copy.save.max, copy.live.max = copy.live.max, 0
    if atomic.CompareAndSwapPointer(&c.state, statePtr, &copy) { break }
}
```

That results in a memory alloc, so not sure about the trade-off.

Reply:
Yeah, I'd like to avoid that. I'm thinking of this aggregator as a good choice for high-concurrency and low-cost. Another aggregator should be chosen if this issue matters.